caffe2/serialize/versions.h

```diff
@@ -78,5 +78,9 @@ static_assert(kProducedBytecodeVersion >= kProducedFileFormatVersion,
 // we should support this model_version. For example, we provide a wrapper to
 // handle an updated operator.
 constexpr uint64_t kMinSupportedBytecodeVersion = 0x3L;
+// Introduce kMinProducedBytecodeVersion for limited forward compatibility
+// support of bytecode. We can support generating model version within
+// [kkMinProducedBytecodeVersion, kProducedFileFormatVersion].
+constexpr uint64_t kMinProducedBytecodeVersion = 0x4L;
```

Comment:
My understanding is that specific bytecode version will allow you to generate model in the format of that version.

I dont remember our stance on forward compatibility, but this seems to add forward compatibility within a range of bytecode versions. 

What are the opinions on how or if to support FC?

Reply:
Yeah I meant to put it for broader visibility. In export function, we can check if they are within the range [kMinProducedBytecodeVersion, kProducedFileFormatVersion], and have a few other const number/string for model v4 and v5. After a second thought, it's probably better to keep the version knowledge in export function, since it's not necessarily needed and need extra effort to maintain it.

Reply:
> My understanding is that specific bytecode version will allow you to generate model in the format of that version.
> 
> I dont remember our stance on forward compatibility, but this seems to add forward compatibility within a range of bytecode versions.
> 
> What are the opinions on how or if to support FC?

@iseeyuan, @raziel and I had a discussion offline in terms of how to enable forward compatibility. Copy Raziel's suggestion here:
> My suggestion is to break this as 2 concerns:
1. Regular version bump flow, which implies handling backwards compatibility for previous versions.
This means that, in your example, the current version (aka kProducedFileFormatVersion) will be bumped from v4 to v5.
And that, as with any version bump, you'll have to decide and, if possible, implement backwards compatibility with the previous version v4.
2. Implement forward compatibility, by means of enabling the new framework code to produce an old model version.
This means that, in your example, you will have to write an "adapter" that allows producing models in the old version (v4). For this you just need the current version (aka kProducedFileFormatVersion) so you know from what version your back-porting from.
Note that by default the framework should produce models in the current version (v5). The above back-porting capability should be a separate API that those other pieces of infra that need to produce old models can use.
This way you don't need a new field, nor the complex and unintuitive logic to bump them separately and keep them consistent.
Instead you just do regular versioning for backwards compatibility, and add the capability of back-port models to older versions to address FC breakages.